Source/JavaScriptCore/interpreter/Interpreter.cpp

```diff
@@ -1005,8 +1003,11 @@ JSValue Interpreter::executeProgram(const SourceCode& source, JSGlobalObject*, J
         {
             CodeBlock* tempCodeBlock;
             program->prepareForExecution<ProgramExecutable>(vm, nullptr, scope, CodeForCall, tempCodeBlock);
-            RETURN_IF_EXCEPTION(throwScope, throwScope.exception());
-
+            ASSERT(!!throwScope.exception() == vm.traps().needHandling(JSC::VMTraps::NeedExceptionHandling));
+            if (UNLIKELY(vm.traps().needHandling(VMTraps::AllEvents))) {
+                if (vm.hasExceptionsAfterHandlingTraps())
+                    return throwScope.exception();
+            }
```

Comment:
This is wrong.  This bypasses the DeferTraps scope.  Instead, let's introduce a new macro as follows and use that:
```
#define RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope__, value__) do { \
        JSC::VM& vm = (scope__).vm(); \
        if (UNLIKELY(vm.exception())) \
            return value__; \
    } while (false)
```

Reply:
Fixed.